[CSL-1373] Tool to analyse a blockchain

adinapoli-iohk opened the pull request:
The long-awaited blockchain tool is here! 🎉 

Hopefully it will be useful to the core team even in this first incarnation. Demo time, first!

## Displaying stats in 'Table' mode
![screen shot 2017-09-05 at 17 38 58](https://user-images.githubusercontent.com/29383371/30070840-0173d8d6-9265-11e7-9ec4-7c3d47190338.png)

## Displaying stats in 'CSV' mode
![screen shot 2017-09-05 at 17 39 16](https://user-images.githubusercontent.com/29383371/30070843-045249c0-9265-11e7-8c13-844dd49b4a56.png)

## Display stats in 'Human' mode
![screen shot 2017-09-05 at 17 39 36](https://user-images.githubusercontent.com/29383371/30070846-0488043e-9265-11e7-80bb-2463b26c1019.png)

We support a number of basic-but-useful options:

```
Cardano SL blockchain generator

Usage: cardano-blockchain-analyser --db FILEPATH ([-k] | [-m] | [-g] | [-b])
                                   [--print-mode [human|csv|table]]
                                   [-i|--incremental]
  Analyse a blockchain and spit out useful metrics.

Available options:
  -h,--help                Show this help text
  --db FILEPATH            Location of the database where the blockchain is
                           stored.
  -k                       Display block counts in kilobytes (KB).
  -m                       Display block counts in megabytes (MB).
  -g                       Display block counts in gigabytes (GB).
  -b                       Display block counts in bytes (B).
  --print-mode [human|csv|table]
                           Select the desidered rendering mode, one between
                           'human', 'csv' or 'table'.
  -i,--incremental         Run in incremental mode. In this mode, table output
                           will be disabled and rendered as a .csv, as is not
                           possible to generate nice-looking tables whilst
                           reading the blockchain one block at time. You almost
                           always want to be using this mode for huge
                           blockchains, as it's much more memory efficient.

Command example:

  cardano-blockchain-analyser --db /path/to/existing/db
```

## Caveats

In order for the tool to be able to lookup blocks & headers is **required** to build it with the same `productionMode` used to build/forge the input database.

int-index commented:
Can you also remove `scripts/analyze/blocks.sh` as it's obsoleted by this new tool?

georgeee commented:
Neat! Screenshots look promising!

gromakovsky commented on tools/src/blockchain-analyser/Main.hs:
```diff
@@ -0,0 +1,98 @@
+-- big DBs might not be feasible.
+analyseBlockchainLazily :: HasCoreConstants => CLIOptions -> HeaderHash -> BlockchainInspector ()
+analyseBlockchainLazily cli currentTip = do
+    allBlocks <- go currentTip mempty
```
There is `loadBlundsFromTipWhile` in `Pos.DB.DB` which does the same thing (except that it returns `NewestFirst`).

adinapoli-iohk replied:
I've tried to use this one, but the `condition` of `loadBlundsFromTipWhile` is a pure function whereas my terminating condition is whether I can fetch the next block or not, for which I need to run a monadic computation in the DB monad. Dunno if there is some other trick I can play, but otherwise I guess we'll have to live with the current implementation 😞

gromakovsky replied:
@adinapoli-iohk just pass `const True`, it will load all blocks and will stop when it reaches the deepest genesis block. That's exactly what you need, AFAIU. A small advantage of using `loadBlundsFromTipWhile` is that there will be `DBMalformed` exception when there is no previous block in DB for any block different from the deepest genesis block (which indeed shouldn't happen for valid DB), while your version will ignore it silently.

gromakovsky commented on tools/src/blockchain-analyser/Rendering.hs:
```diff
@@ -0,0 +1,197 @@
+                 , renderBlock
+                 , renderBlocks
+                 , renderHeader
+                 ) where
```
The way we format export lists is defined [here](https://github.com/input-output-hk/cardano-sl-style-guides/blob/master/haskell-style-guide.md#export-lists).
I know it's minor, but I want it to be consistent with other code. Note that you module `Options` follows that rule, so even in this PR formatting is different. When it's different, some people will change it, it will lead to unnecessary extra diff and noise, sometimes even conflicts.

adinapoli-iohk commented on tools/src/blockchain-analyser/Main.hs:
```diff
@@ -0,0 +1,98 @@
+-- | Analyse the blockchain, printing useful statistics.
+analyseBlockchain :: HasCoreConstants => CLIOptions -> HeaderHash -> BlockchainInspector ()
+analyseBlockchain cli tip =
+    if incremental cli then do liftIO $ putText (renderHeader cli)
```
Oh, neat, I assumed it was not 😉

adinapoli-iohk commented on tools/src/blockchain-analyser/Types.hs:
```diff
@@ -0,0 +1,84 @@
+-- | Enough context for analysing blocks.
+data BlockchainInspectorContext = BlockchainInspectorContext
+    { tbgcDBSum   :: DB.DBSum
+    , tbgcNodeDBs :: DB.NodeDBs
```
I think you are right -- Initially I wasn't aware of the difference between the different modes so I wanted to have `DBSum` at hand in order to have a bunch of instances defined, but I don't think this is necessary anymore.

adinapoli-iohk commented on tools/src/blockchain-analyser/Rendering.hs:
```diff
@@ -0,0 +1,197 @@
+        blockType     = either (const "GENESIS") (const "MAIN") block
+        slot          = maybe mempty (sformat build . getSlotIndex . siSlot) (getSlot blockHeader)
+        leader        = maybe mempty (sformat build) (getLeader blockHeader)
+        txCount       = sformat build (length (getTxs block))
```
Neat!

adinapoli-iohk commented:
@gromakovsky Ok, I think I have addressed most of the feedback (module the one on `allBlock` which I think we might need to stick to the status quo). Let me know what you think!

gromakovsky commented on tools/src/blockchain-analyser/Types.hs:
```diff
@@ -0,0 +1,74 @@
+    dbGetHeader = BDB.dbGetHeaderDefault @SscGodTossing
+
+prevBlock :: HasCoreConstants => Block SscGodTossing -> HeaderHash
+prevBlock = either (view prevBlockL) (view prevBlockL)
```
Actually there is
```
instance (HasPrevBlock s, HasPrevBlock s') =>
         HasPrevBlock (Either s s') where
    prevBlockL = choosing prevBlockL prevBlockL
```
so you can write `view prevBlockL` (no `either`).

gromakovsky commented on tools/src/blockchain-analyser/Types.hs:
```diff
@@ -0,0 +1,74 @@
+type DBFolderStat = (Text, Integer)
+
+-- | Enough context for analysing blocks.
+data BlockchainInspectorContext = BlockchainInspectorContext { tbgcNodeDBs :: DB.NodeDBs }
```
I just noticed: `tbgc` prefix looks strange, usually we use abbreviation of type name as a prefix (so `bic` in this case).

adinapoli-iohk commented:
@gromakovsky ok, it should all be addressed now! Note how I refrained from using `Blund` inside `Rendering` so the `analyseBlockchainLazily` still needs to fetch each `Undo` manually and the lookup can fail with `Nothing`. We could in principle use one of the function which makes the tool blowup with a `MalformedDB` exception, but I think that for tools like this it's nice to have a bit of slack and be more lenient than usual.
